src/exec/engine.go

```diff
@@ -123,53 +96,41 @@ func (e Engine) acquireConfig() (cfg config.Config, err error) {
 	return
 }
 
-// fetchConfig returns the configuration from the first available provider or
-// returns an error if none of the providers are available.
-func fetchConfig(providers []providers.Provider, timeout time.Duration) (config.Config, error) {
-	if provider, err := selectProvider(providers, timeout); err == nil {
+// fetchConfig returns the configuration from the provider or returns an error
+// if the provider is unavailable.
+func fetchConfig(provider providers.Provider, timeout time.Duration) (config.Config, error) {
+	if err := waitForProvider(provider, timeout); err == nil {
 		return provider.FetchConfig()
 	} else {
 		return config.Config{}, err
 	}
 }
 
-// selectProvider chooses the first online provider, given a list of providers
-// and a timeout. If none of the providers will ever be online, or if the
-// timeout elapses before any providers are online, this returns an appropriate
-// error.
-func selectProvider(ps []providers.Provider, timeout time.Duration) (providers.Provider, error) {
-	online := make(chan providers.Provider)
-	wg := sync.WaitGroup{}
+// waitForProvider waits for the provider to come online. If the provider will
+// never be online, or if the timeout elapses before it is online, this returns
+// an appropriate error.
+func waitForProvider(provider providers.Provider, timeout time.Duration) error {
+	online := make(chan bool, 1)
 	stop := make(chan struct{})
 	defer close(stop)
 
-	for _, p := range ps {
-		wg.Add(1)
-		go func(provider providers.Provider) {
-			defer wg.Done()
-
-			for {
-				if provider.IsOnline() {
-					online <- provider
-					return
-				} else if !provider.ShouldRetry() {
-					return
-				}
-
-				select {
-				case <-time.After(provider.BackoffDuration()):
-				case <-stop:
-					return
-				}
+	go func(provider providers.Provider) {
```

Comment:
Why even pass in the provider when the closure can just access the provider from the calling scope?  It made sense in the old code, since there was a list of providers and the closure was acting on one at a time... but it seems like now this would be less verbose to just drop the argument from the closure.